### ISSUE TITLE
Add support for MI Air Conditioner Companion MCN

### ISSUE THREAD

**Issue by jianyun8023** [merdok/homebridge-miot#52]
Hello,

Would it be possible to add support for the Mi Air Conditioner Companion MCN?

**Model:** `lumi.acpartner.mcn02`
**SKU:** `KTBL03LM`
**Link:** https://www.mi.com/buy/detail?product_id=9726

When setup in the plugin, it seems to connect successfully and is seen as a switch, but there are no additional controls. 
I see that [python-miio](https://github.com/rytilahti/python-miio) has his support, you can refer to [python-miio#809](https://github.com/rytilahti/python-miio/pull/809).
If there is anything I can provide (logs, testing, etc.), I would be happy to help the best I can.

**Comment by merdok** [comment]
Hi,

I am wondering myself how this actually works that outlets can be air conditioners?😅

Anyway, yes adding support for the device should be straight forward, the only thing which we will need to find out if this is actually a miio or miot device. From the issue you linked I saw that they use miio commands. When you right now add the device when you try to use the switch, does it actually do anything? Could you post debug log when you use the switch so we can check out if the device accepts the command?

If the device does not accept miot commands locally then I can use the MiCloud to make it work. So we just need to figure out if the device can be controlled locally using miot or if it requires MiCloud...

**Comment by jianyun8023** [comment]
The air conditioner on the socket, I researched some time ago, it should be a miot device.

https://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:air-conditioner:0000A004:lumi-mcn02:1

The following is the desensitization log of homebridge:
[hb.log](https://github.com/merdok/homebridge-miot/files/6749074/hb.log)

**Comment by merdok** [comment]
What i can see from the log is that the property fetch call times out and this is a strong indicator that the device is not a miot device. Older devices use  the old miio protocol but xiaomi built in a translator in the cloud to convert the miot commands from the specs which you posted here to local miio commands. So yes, i can add support for the device but it will require a MiCloud connection. You can also already check out now if the device will work with MiCloud by adding the following lines to your device config in the config.json:

```
"micloud":
 {
          "username": "email",
          "password": "pwd",
          "country": "cn",
         "forceMiCloud": true
},
```

Add your username and password and also the cloud country where your devices are located. That should make the device successfully sync the power property. Let me know if that is the case, and then i will add it.

**Comment by jianyun8023** [comment]
Use micloud to successfully control the power switch.

```
[7/2/2021, 12:36:20 AM] [miot] [lumi.acpartner.mcn02] Splitting properties into chunks. Number of chunks: 1. Chunk size: 14
[7/2/2021, 12:36:20 AM] [miot] [lumi.acpartner.mcn02] Chunks:  [
 [
  "power"
 ]
]
[7/2/2021, 12:36:20 AM] [miot] [lumi.acpartner.mcn02] MiCloud request https://api.io.mi.com/app/miotspec/prop/get - {"_nonce":"oFnyPHZXxsIBnUyE","data":"{\"params\":[{\"did\":\"333641900\",\"siid\":2,\"piid\":1}]}","signature":"Sg+tJCHFHghc/v4MAw5b+PmXf8in3bV5850z+3/2MBA="}
[7/2/2021, 12:36:20 AM] [miot] [lumi.acpartner.mcn02] Device properties updated: 
 [
  {
    "name": "power",
    "value": true
  }
]

[7/2/2021, 12:38:01 AM] [miot] [lumi.acpartner.mcn02] Set power property request! RAW: {"did":"333641900","siid":2,"piid":1,"value":false}
[7/2/2021, 12:38:01 AM] [miot] [lumi.acpartner.mcn02] MiCloud request https://api.io.mi.com/app/miotspec/prop/set - {"_nonce":"fksI9hG/ZPsBnUyG","data":"{\"params\":[{\"did\":\"333641900\",\"siid\":2,\"piid\":1,\"value\":false}]}","signature":"ja9ky0UA/1LBaw0TFzb6iZYspc/9OyxrTl1XaCtDDYE="}
[7/2/2021, 12:38:01 AM] [miot] [lumi.acpartner.mcn02] Successfully set property power to value false! Response: [{"did":"333641900","siid":2,"piid":1,"code":0}]
[7/2/2021, 12:38:02 AM] [miot] [lumi.acpartner.mcn02] Set power property request! RAW: {"did":"333641900","siid":2,"piid":1,"value":true}
[7/2/2021, 12:38:02 AM] [miot] [lumi.acpartner.mcn02] MiCloud request https://api.io.mi.com/app/miotspec/prop/set - {"_nonce":"M2r43PuZa+MBnUyG","data":"{\"params\":[{\"did\":\"333641900\",\"siid\":2,\"piid\":1,\"value\":true}]}","signature":"F435MotsgSn+t/+VryCDV9bgz5QpS6y/PwSuS6W+wxo="}
[7/2/2021, 12:38:02 AM] [miot] [lumi.acpartner.mcn02] Successfully set property power to value true! Response: [{"did":"333641900","siid":2,"piid":1,"code":0}]

```
![image](https://user-images.githubusercontent.com/18240360/124159079-4d011700-dacd-11eb-808c-8d270ce72568.png)

**Comment by merdok** [comment]
Perfect! So as expected this is a miio device. I will add full support for it in the next update! Will let you know when the update is ready!

**Comment by jianyun8023** [comment]
@merdok Wow, that’s great~ Thank you very much.

**Comment by merdok** [comment]
Version 0.9.7 is now online with support for the device. Please try it and let me know!

**Comment by ahmadnazirafiq** [comment]
Hi,

Works great. However, when trying to tap and hold the tile in the Home app, it defaults to OFF thus turning it back off everytime I wanted to adjust the temperature. Also, the temperature meter stated 0.0 all the time.

![IMG_9564](https://user-images.githubusercontent.com/66505845/126053421-b0d136b8-9ccf-4cc1-8b7d-d7016e20f1db.jpg)

**Comment by merdok** [comment]
I think you do not have the latest version of the plugin hece there is the the error in the console. Please update to the latest version. 

The temperature will be reported as 0.0 by the accessory since the device does not support temperature reporting

**Comment by ahmadnazirafiq** [comment]
Hi,

Appreciate your response. The plugin runs on the latest version tho, which is 0.9.7.

As for the temperature, maybe if it could follow the temperature we manually adjust in the tile? I’m using a separate “Mi Heatercooler” plugin for the the old version of Mi Air Conditioner Controller and it does that.

**Comment by merdok** [comment]
Ah, you are right, it will be fixed in the next update. Also after the update the tile will show the temperature which you set.

**Comment by ahmadnazirafiq** [comment]
Thank you! You're awesome!

**Comment by merdok** [comment]
Should now be fixed in version 0.9.8

**Comment by ahmadnazirafiq** [comment]
Hi,

I can confirmed that its fixed now. Works really well. Thank you!
